pose_evaluation/metrics/embedding_distance_metric.py

```diff
@@ -1,104 +1,170 @@
+        if hypothesis is None or reference is None:
+            raise ValueError("Neither 'hypothesis' nor 'reference' can be None.")
 
-        raise ValueError(f"Unsupported distance metric: {self.kind}")
+        try:
+            hypothesis = self._to_batch_tensor_on_device(hypothesis)
+            reference = self._to_batch_tensor_on_device(reference)
+        except RuntimeError as e:
+            raise TypeError(f"Inputs must support conversion to device tensors: {e}") from e
+        return self.score_all(hypothesis, reference).item()
 
     def score_all(
         self,
-        hypotheses: List[Union[np.ndarray, torch.Tensor]],
-        references: List[Union[np.ndarray, torch.Tensor]],
+        hypotheses: List[list | np.ndarray | Tensor],
+        references: List[list | np.ndarray | Tensor],
         progress_bar: bool = True,
-    ) -> torch.Tensor:
+    ) -> Tensor:
         """
-        Compute the pairwise distance between all hypotheses and references. Expects 2D inputs.
+        Compute the pairwise distance between all hypotheses and references.
+        Expects 2D inputs, where each element in the second dimension is one embedding
 
         Args:
-            hypotheses (list[np.ndarray | torch.Tensor]): List of hypothesis embeddings.
-            references (list[np.ndarray | torch.Tensor]): List of reference embeddings.
+            hypotheses (list[list| np.ndarray | Tensor]): List of hypothesis embeddings.
+            references (list[list| np.ndarray | Tensor]): List of reference embeddings.
             progress_bar (bool): Whether to display a progress bar.
 
         Returns:
-            torch.Tensor, distance matrix. Row i is the distances of hypotheses[i] to all rows of references
+            Tensor, distance matrix. Row i is the distances of hypotheses[i] to all rows of references
         """
         # Convert inputs to tensors and stack
-        hypotheses = torch.stack([self._to_tensor(h) for h in hypotheses])
-        references = torch.stack([self._to_tensor(r) for r in references])
-
-        if self.kind == "cosine":
-            # Normalize the tensors along the feature dimension (dim=1)
-            normalized_hypotheses = F.normalize(hypotheses, dim=1)
-            normalized_references = F.normalize(references, dim=1)
-
-            # Calculate cosine similarity between all hypothesis-reference pairs
-            cosine_similarities = torch.matmul(normalized_hypotheses, normalized_references.T)
-
-            # Convert cosine similarities to cosine distances
-            distance_matrix = 1 - cosine_similarities
-        elif self.kind == "l2":
-            # Use broadcasting to calculate pairwise L2 distances
-            diff = hypotheses[:, None, :] - references[None, :, :]
-            distance_matrix = torch.norm(diff, dim=2)
+        hypotheses = torch.stack([self._to_device_tensor(h) for h in hypotheses])
+        references = torch.stack([self._to_device_tensor(r) for r in references])
```

Comment:
these can be replaced with `torch.Tensor(hypotheses).to(this.device)`
instead of converting each to a tensor, you can convert all at once

Reply:
Good idea!

Reply:
Tried it out real quick, it triggered a warning: "<ipython-input-3-858322941910>:7: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:278.)"

Lemme dig a bit more

Reply:
https://github.com/pytorch/pytorch/issues/13918

A long-known issue, apparently. It seems the best practice currently is to  do 
```
np_array = np.array(list_of_ndarrays)
tensor = torch.from_numpy(np_array)
```

Reply:
Did some checking. For 30k embeddings of size 768, the difference is quite large, as in several seconds vs 7 milliseconds. 

```
import torch 
import numpy as np
import time
from tqdm.notebook import tqdm

if torch.cuda.is_available():
  torch.set_default_device('cuda')
else:
  torch.set_default_device('cpu')
# create a list of numpy arrays of length 768
def convert_list_of_np_arrays_to_one_big_tensor_via_stacking(np_array_list):
  torch_tensors = [torch.tensor(np_tensor) for np_tensor in np_array_list]
  stacked_torch_tensors = torch.stack(torch_tensors)
  return stacked_torch_tensors

# def convert_list_of_np_arrays_to_one_big_tensor_via_concat(np_array_list):
#   torch_tensors = [torch.tensor(np_tensor) for np_tensor in np_array_list]
#   concatenated_torch_tensors = torch.cat(torch_tensors, dim=0)
#   return concatenated_torch_tensors # wrong shape, one dimension 

def convert_list_of_np_arrays_to_one_big_tensor_via_concat(np_array_list):
  torch_tensors = [torch.tensor(np_tensor).unsqueeze(0) for np_tensor in np_array_list]  # Add unsqueeze(0)
  concatenated_torch_tensors = torch.cat(torch_tensors, dim=0)
  return concatenated_torch_tensors # somehow this results in a bit of numerical precision differences, where the others don't

def convert_list_of_np_arrays_to_one_big_tensor_via_tensor_constructor(np_array_list):
  return torch.Tensor(np_array_list)

def convert_list_of_np_arrays_to_one_big_tensor_via_np_array_and_from_numpy(np_array_list):
  np_array = np.array(np_array_list)
  return torch.from_numpy(np_array)

def benchmark_np_tensors(trials =10, embedding_length=768, list_length=1000):
  print("Benchmarking list of np arrays to Tensor")

  conversion_methods_dispatch_table = {
    "stacking": convert_list_of_np_arrays_to_one_big_tensor_via_stacking,
    "concat": convert_list_of_np_arrays_to_one_big_tensor_via_concat, # fails the assertion.
    "tensor_constructor": convert_list_of_np_arrays_to_one_big_tensor_via_tensor_constructor,
    "np_array_and_from_numpy": convert_list_of_np_arrays_to_one_big_tensor_via_np_array_and_from_numpy,
  }

  durations = {}
    

  for conversion_method_desc in conversion_methods_dispatch_table:
    method = conversion_methods_dispatch_table[conversion_method_desc]
    print(f"Benchmarking {method.__name__}, using device {torch.get_default_device()}")
    durations[conversion_method_desc] = []
    try: 
      for i in tqdm(range(trials)):
        np_array_list = [np.random.rand(embedding_length) for _ in range(list_length)]        

        start_time = time.perf_counter()
        tensor = method(np_array_list)
        end_time = time.perf_counter()
        duration = end_time - start_time
        durations[conversion_method_desc].append(duration)

        tensor = tensor.to('cpu')

        torch.cuda.empty_cache()
        
        # check the ndims are correct, and shape is correct
        assert tensor.ndim == 2, f"Tensor has the wrong dimensions! ndim: {tensor.ndim}, shape: {tensor.shape}"
        assert tensor.shape == (list_length, embedding_length), f"Tensor has the wrong shape! ndim: {tensor.ndim}, shape: {tensor.shape}"

        # check if the first element of every array is equal to the first element of every tensor
        for j in range(list_length):
          if not np.allclose(np_array_list[j][0], tensor[j][0].numpy(), atol=1e-6):
            raise AssertionError(f"First element of array {j} and tensor {j} do not match: \n*\t{np_array_list[j][0]} vs \n*\t{tensor[j][0].numpy()}")

        # check if the first array is equal to the first tensor
        if not np.allclose(np_array_list[0], tensor[0].numpy(), atol=1e-6):
          raise AssertionError(f"First array and tensor do not match: \n*\t{np_array_list[0]} vs \n*\t{tensor[0].numpy()}")

        
    except AssertionError as e:
      print(f"AssertionError: {e}")



      
  # print averages
  for conversion_method_desc in conversion_methods_dispatch_table:
    print(f"{conversion_method_desc}: {np.mean(durations[conversion_method_desc]):.4f} seconds")

benchmark_np_tensors(3, 768, 30000)
```

Reply:
Results are the same if I change the default device to GPU

Reply:
The next issue is: what method will work for all inputs?
Possible inputs could be: 
* a List of ndarrays. we need to do np_array = np.array(list_of_ndarrays)
* A Tensor -> if we call np.array on this we get an error
* A List of Tensors, if we call np.array on this we get an error

Reply:
in this case, make a method `embedding_to_tensor(embedding: Embedding)` and checks what needs to be done. I think that would be best

Reply:
That's probably wise, I am sometimes tempted to premature optimization, haha

Reply:
Added some checks in the _to_batch_tensor_on_device function